### ISSUE TITLE
Add timeout implementation to Deferred, based on cancellation support

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5786]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5786|
|Type|enhancement|
|Created|2012-07-16 00:26:44Z|
|Branch|https://github.com/twisted/twisted/tree/timeout-deferreds-5786-5|

If you can cancel a Deferred, you should be able to time it out as well.

Some non-obvious points:

  1. It'd be useful to have an ability to distinguish between different timeouts (and cancellation) via the exception in the Failure.
  1. You may want to add multiple timeouts, and have them affect different levels of the callback chain. Consider a high-level request than opens a connection, then sends out a HTTP request. The code that creates the network connection might want to add a 60 second timeout on the connection attempt. Code calling the combo high-level API might want a 10 minute timeout, which applies to combination of network connection and HTTP request.

<details><summary>Searchable metadata</summary>

```
trac-id__5786 5786
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__timeout_deferreds_5786_5 timeout-deferreds-5786-5
branch_author__itamarst__itamar__kaizhang itamarst, itamar, kaizhang
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1342398404000000 1342398404000000
changetime__1475245475991716 1475245475991716
version__None None
owner__Mark_Williams__markrwilliams_____ Mark Williams <markrwilliams@...>
cc__twm cc__mithrandi@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Replying to [itamar](#5786):
> If you can cancel a Deferred, you should be able to time it out as well.
> 
> Some non-obvious points:

These _are_ non-obvious :).

>   1. It'd be useful to have a TimeoutError, separate from CancelledError.

Counterpoint: no, it wouldn't.  What code would ever want to know this, and why?  Cancellation can occur for any number of causes; timeouts are one of them, but there are plenty of others, and I think it would actually be bad to have code distinguish, for the most part, because you don't want code that works when a user explicitly cancels but fails when a timeout happens.

>   2. You may want to add multiple timeouts, and have them affect different levels of the callback chain. Consider a high-level request than opens a connection, then sends out a HTTP request. The code that creates the network connection might want to add a 60 second timeout on the connection attempt. Code calling the combo high-level API might want a 10 minute timeout, which applies to combination of network connection and HTTP request.

We should probably have a discussion about how Tubes handle this case (the "progress" callback), since there is _some_ stuff that you can do that's generic in support of this use-case, but I don't think it fits at this level.

Having a `Deferred` publish any API to set timeouts on events, e.g. other `Deferreds` and function calls, that aren't in your code and you can't observe, is probably bad.  These would just be arguments to the functions that initiate certain activities, and those timeouts would be handled internally and would have specific error conditions associated with them.  For example, you will also want a timeout on "time between bytes received on this connection" when a download is in progress, as distinct from "time to download the whole file" and "time to initially connect", but that's none of the `Deferred`'s business (how would you even expose all of those distinct stages?).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Basically I disagree with this feature and I would like to reject it and close the ticket as invalid.  I was super happy when `setTimeout` went away, and I want to get `defer` out of `twisted.internet`.  If we do need such a feature I really think it should be external to Deferred itself, but I would rather just have `reactor.callLater(n, d.cancel)` be the idiom here.

But without a lot more justification in terms of concrete use-cases, I would say we just shouldn't do this.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Point 2 as a use case was based on actual code I've just written.

The implementation I will provide is indeed just a slight improvement over `reactor.callLater(n, d.cancel)` (which is insufficient only because it leaves dangling DelayedCalls). The minimal implementation you suggest, or the slightly improved one I intend to implement, both provide point 2 as a side effect of the implementation:

```
def getConnection(reactor, endpoint):
    d = endpoint.connect(Factory())
    addTimeout(reactor, d, 60) # 1 minute timeout limited to connection
    return d

def sendCommand(reactor, command):
    d = getConnection(reactor, makeEndpoint())
    d.addCallback(lambda conn: conn.request(command))
    addTimeout(reactor, d, 600) # 10 minute timeout on whole process
    return d
```

Even if your implementation was sufficient (and it almost is), it should be documented, and the nuances in point 2 explained.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

A correction: your naive implementation (in addition the resource leak) does *not* enable to functionality of multiple timeouts as in the example, because it doesn't cancel the first timeout when the Deferred fires.

So another reason to actually implement this and provide it in Twisted: the naive way is broken.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

To expand further on the use case - sending a query involves:

1. Getting a connection from a connection pool, or if none are available making a new one.
2. Sending a query, returning response.

You really want to support different timeouts for those two, even though it's the same Deferred object. But because the Deferred is waiting for the 2nd action, which returned its own Deferred, if you don't cancel the timeout for the connection when it succeeds, it will cancel the query even though the query still has a whole bunch more time left.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

As for point 1, whose description I just updated:

Given multiple timeouts on a single Deferred, you want a way for a GUI, or a log message, to distinguish between them. Did the operation fail because the connection attempt timed out or because the command timed out? Now, you could do this yourself; every time you call addTimeout you also do:

```
addTimeout(reactor, d, 60)
def explainReason(reason):
    reason.trap(CancelledError)
    return Failure(MyTimeoutReason())
d.addErrback(explainReason)
```

But of course that will give the wrong result if someone *manually* cancels, and it's also needlessly repetitive boiler plate.

So point 1 is also necessary.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_5):
> To expand further on the use case - sending a query involves:
> 
> 1. Getting a connection from a connection pool, or if none are available making a new one.
> 2. Sending a query, returning response.
> 
> You really want to support different timeouts for those two, even though it's the same Deferred object. But because the Deferred is waiting for the 2nd action, which returned its own Deferred, if you don't cancel the timeout for the connection when it succeeds, it will cancel the query even though the query still has a whole bunch more time left.

What API do you propose for adjusting these timeouts independently?  Or even propagating their existence up through multiple layers of the Deferred stack?  A Deferred can have any number of callbacks on it, and in some cases you don't even know in advance (think HTTP redirects).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_6):
> As for point 1, whose description I just updated:
> 
> Given multiple timeouts on a single Deferred, you want a way for a GUI, or a log message, to distinguish between them.

OK, that might need to be a feature built in to deferred cancellation; I'd rather that it not be reflected by the type of the exception, but with an attribute of some kind.  But, it's not deterministic. You cancel the Deferred from the consumer's perspective, and the originator gets to callback or errback it with whatever value or error it likes.  You can't depend on the ability to catch different errors.

> Did the operation fail because the connection attempt timed out or because the command timed out? Now, you could do this yourself; every time you call addTimeout you also do:

So wait - are you saying that the timeout will be creating a _new_ Deferred for each timeout, and ... chaining the previous Deferred in?  That's interesting.  That _could_ reliably give you different exceptions, by having a custom canceler.  I hadn't thought of that.

> But of course that will give the wrong result if someone *manually* cancels, and it's also needlessly repetitive boiler plate.
> 
> So point 1 is also necessary.

I still don't quite see how this follows, but maybe I'm starting to get your point.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

> What API do you propose for adjusting these timeouts independently?  Or even propagating their existence up through multiple layers of the Deferred stack?  A Deferred can have any number of callbacks on it, and in some cases you don't even know in advance (think HTTP redirects).

Hmm. I hadn't thought much about adjustment; I guess my thought is that Deferred timeouts are for the absolute it-must-take-this-long-and-no-longer sort of timeout, where you set it and forget it. As opposed to "did I get any bytes in past 60 seconds" where you're constantly resetting, but that happens on the protocol level. In particular, because Deferreds are a "give me one result" API.

Does this make sense? Or can you think of a counter-example where you would want to adjust it? (Perhaps each layer adding a timeout should be able to adjust its timeout, though, even if there's no way to access other layers' timeouts).

Propagation happens automatically via the way cancellation knows how to pass cancellation to Deferreds returned by callbacks.

At this point I think I'll switch to coding and tests, so we have something more concrete to discuss.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

(In [34883]) Branching to 'deferred-timeouts-5786'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

(In [34886]) Custom exceptions. Refs [#5786](https://github.com/twisted/twisted/issues/5786).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince commented|
|-|-|

If adjusting the timeout is a usecase you want to support, it seems like the easiest way to handle that is to have setTimeout return an object that can adjust that timeout. Just returning the ```IDelayedCall}} would perhaps be sufficient```{IDelayedCall}} would perhaps be sufficient.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

The basic implementation is done (other than what Tom suggested), so I'd appreciate comments before moving on to documentation.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst commented|
|-|-|

The implementation is completely done. I'd appreciate a review on the design before moving on to documentation.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Reviewing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set owner to @itamarst<br>@glyph set status to `new`|
|-|-|

Hey Itamar.  Thanks for pursuing this.

Aesthetic judgments:

    1. The naming is just awful.
        1. The 'add' in the name `addDeferredTimeout` implies that this is something you might want to use multiple times on the same Deferred.  But that would result in multiple cancellations.  Which, of course, you might want in some very rarified circumestances, but generally that would not be the case.  I suggest a verb-ier name, like `timeOutDeferred`.
        2. The `callID` variable implies that `callLater` returns some opaque identifier.  It doesn't, and it hasn't fora really long time.  The thing it does return, an `IDelayedCall`, would be better encapsulated by a name like, e.g. `delayedCall`.
        3. In naming the `exception` parameter, you should describe its role and not its type.  What is this exception used for?
    2. The documentation is basically incoherent.
        1. The word 'timeout' is used 11 times in the documentation.  First as a noun, in the name, where it's slightly ambiguous what it's describing.  Then it's used as a noun describing something else - the amount of time before the timeout.  Then, as the first word, it's used as a verb.  Please clean this up to use "timeout" as little as possible and instead use active words that describe what's happening, like "canceled" or "called back" or "failed".  I recommend the parameter be called something like "seconds" or "interval" or "delay".
        2. "Registering a custom canceller is recommended."  Recommended by whom?  For what purpose?  In what way?  (In other words: avoid the passive voice, please.)  What would such a custom canceller do, especially with respect to this specific API which recommends it?  Also: you don't ''register'' a custom canceller, you ''specify'' one.  Registration implies a post-creation API that modifies some state.  Deferreds must be created with their canceller.  Furthermore, the user of the `addDeferredTimeout` API will, often as not, be consuming rather than originating a `Deferred`, and therefore won't be _able_ to "register" a custom canceller.  I don't think that this docstring should recommend anything specific about cancellers, custom or otherwise, but it should provide a reference to the cancellation documentation (although something as terse as `@see: L{Deferred.cancel}` would be sufficient.)
        3. I do not care about the API's inner thoughts and feelings about its role or its level of knowledge.  I want to know what it does.  The whole second paragraph that describes whether it is "aware" of certain callbacks or not is very confusing.  Please rephrase it to describe its behavior with respect to those callbacks - will they be called?
        4. NO AMBIGUOUS ANTECEDENTS.  "it will not be able to time it out". WHAT.
        5. Omit the needless word.  "In particular" seems to convey no information.  "typically the reactor... for testing" seems similarly useless.  If this is important information, it should be on `IReactorTime` itself, not on some random API.
    3. The test docstrings use unclear language to describe what they're testing.
        1. `test_noTimeoutIfCallback` - says that it "will not cause a timeout".  This is a negative assertion and therefore impossible to prove :).  It should say something more specific about the state of the Deferred.  Several other tests refer to 'causing timeouts' as well.
        2. `test_timeout` says something about the `Deferred` being "fired in time".  In what time?
        3. I am a little confused by what it means to cancel the timeout "manually" as opposed to just preventing the Deferred from being cancelled.
    4. Attempting to throw an exception over the heads of a bunch of intervening callbacks, and hoping that it will reach the target point in the callback stack, strikes me as poor form.  This should be introduced at some point in the call chain where you can make a definitive statement about all the enumerated failure modes.  You need to be able to specify not only the exception that you want to ''raise'', but also the exception(s) that you want to ''trap''.  But, ultimately, I don't think that this feature of cancellation is really helpful at all and I think it should be removed.
    5. A more useful feature, which I don't see any way to easily add to this externally, would be to force the timeout to happen immediately
    6. `Failure.check` is more efficient than `Failure.trap` and would likely read better in this code anyway.

Mandatory Stuff (violated policies, et. al.):

    1. The docstring is missing a couple of required elements.
        1. It's missing a `@return` and it doesn't return `None`.  Documentation of the return type is always super important.
        2. You should also include separate `@type` annotations for each parameter (and an `@rtype`).
    2. This needs a place in the narrative documentation, as your pre-review comment suggests.  A narrative explanation might help to better explain the relationship between the call chain and the timeout, which, except for the exception-type-translation stuff (which I don't think works entirely right and you don't really need anyway) is actually quite straightforward.
    3. The exception type translation suffers from a fatal flaw and can never really do what it's advertising 100% reliably.  If you have a deferred A whose callback returns a deferred B, and B's custom canceller does a partial recovery and returns a result, but then the ''next'' callback on A raises `CancelledError` for some unrelated reason, it will appear as though the timeout exception was raised, when in fact, the timeout just made the operation complete faster and the cancellation was from somewhere else.  But, again, I don't think this feature serves any real purpose: the _real_ utility in something like this would be in conjunction with a custom canceller that would make the _first_ exception raised by the `Deferred` being cancelled into a `TimeoutError`.  More importantly though, nothing about the documentation or the tests explains why I would care about the distinction between these exceptions at the consumer's point in the call chain.  This needs to be explained, and explained in fairly painstaking detail.

In short, this appears to be almost-adequately documented and adequately tested, and I can see no procedural reason not to merge it.  However, I'm fairly unhappy with the design and the documentation seems as unhelpful as some of our oldest, crummiest documentation.  I think this would be a lot better if it didn't attempt the exception-translation stuff and just did the one thing that users would generally forget, which is the reference-cleanup of the timeout callback itself when the `Deferred` fires; exception translation could be added as a separate feature later.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_19"></a>|@itamarst commented|
|-|-|

Most of this is things I can just go and fix, but I don't want to punt on the custom exception use case without having a design. And it is a real use case: if my HTTP request timed out, "failed to look up host" is a very different thing than "response timed out", if only for debugging purposes (but possibly for business logic as well).

A reasonable design, which is implicit in your critique, is having the custom canceller be able to raise some exception other than CancelledError... but that is not currently part of the cancellation API. Does that seem like a reasonable feature to add? If so, it could then be used to implement custom exceptions for timeouts. Obviously these would be two additional, separate tickets.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_20"></a>|@itamarst commented|
|-|-|

I should note, by the way, that in my opinion there is no reason for a "passthrough" errback to trap a CancelledError under any circumstances - CancelledError should only be handled by the final errback. But perhaps there is a use case for that which escapes me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

> And it is a real use case: if my HTTP request timed out, "failed to look up host" is a very different thing than "response timed out", if only for debugging purposes (but possibly for business logic as well).

Have you implemented this using this cancellation API?  It'd be helpful to see the code.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_20):
> I should note, by the way, that in my opinion there is no reason for a "passthrough" errback to trap a CancelledError under any circumstances - CancelledError should only be handled by the final errback. But perhaps there is a use case for that which escapes me.

There's no such thing as a "final" errback, as you well know.  The use-case is that you're using Twisted's APIs, and they provide a certain working model (i.e. Deferred control flow) and you expect it to be honored consistently ;).

More seriously, cancellation would be intercepted for the same reason that any application would implement a `SIGINT` handler.  There are some applications which you just want to do as much work as possible and when a user calls `cancel()` that's their way of saying "OK, I've waited long enough, just give me what you've got".  Consider a 3D rendering program with a preview function, where the output of the Deferred is going to be an image either way, but may be lower resolution if you cancel it before it's all the way done.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_23"></a>|@itamarst set owner to kaizhang|
|-|-|

The work to be done here:

 1. Remove the `exception` argument to the new function. We can always add that functionality later.
 2. Then, address all review comments that are still applicable.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|kaizhang commented|
|-|-|

(In [39637]) Branching to 'deferred-timeouts-5786-2'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_25"></a>|kaizhang removed owner|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deferred-timeouts-5786-2)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince set owner to kaizhang|
|-|-|

Thanks for working on this.

 1. Most of the tests want to be changed to use `successResultOf` and/or `failureResultOf`.
 2. Consider using `IDelayedCall.active` instead of `IReactorTime.getDelayedCalls` to check if the timeout has been cancelled.
 3. In `test_timeout`, the call will trigger after advancing excatly 10.
 4. `test_callbackStack` should probably be split into (at least) two tests, one checking that the cancellation doesn't happen for the first timeout. And the scond, that the cancellation doest happen for the second timeout.

Please resubmit for review after addressing the above points.

 5. It is unclear to me why the final example in the narrative documentation is in the form of interaction with the python interpreter. Condsider changing it to match the other examples.
 6. The pargraph about waiting on a deferred in the documentation for `timeOutDeferred` is awkward, but I don't have any concrete suggetions for improving it. Particularly "The waited deferred" feels ungrammatical.
 7. Consider renaming the function `timeoutDeferred`. I'm inclined to treat "timeout" as a single compound word, rather than two words. Existing usage in twisted seems to agree: 56 files where it is treated as a single word compared to 18).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_28"></a>|kaizhang commented|
|-|-|

Replying to [itamar](#note_27):

Hi itamar, are you going to take over and finish this ticket? I'm sorry that I'm busying with job finding and finishing my paper so I haven't done my tickets yet. It's great if you finish this ticket. Also my vacation will start at next week and last a whole month so I will be available then.

Again, sorry about the delayed tickets.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_29"></a>|@itamarst commented|
|-|-|

I was hoping to finish it, but you can feel free to take it back over if you have the time. It'd be great to have you working on Twisted again if you do!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_30"></a>|@itamarst commented|
|-|-|

(In [41623]) Branching to 'deferred-timeouts-5786-3'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_31"></a>|@itamarst commented|
|-|-|

Still remaining  is fixing up howto, and then it's ready for review again.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_32"></a>|@itamarst commented|
|-|-|

(In [42312]) Branching to 'deferred-timeouts-5786-4'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_33"></a>|@itamarst removed owner|
|-|-|

OK, ready for review again.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_34"></a>|@wallrj set owner to @itamarst|
|-|-|

Thanks Itamar, (and kaizhang)

Notes:
 * Builds pass (apart from some spurious errors)
 * The new documentation renders correctly in sphinx.
 * Tests pass locally
 * Merges cleanly

Points:
 1. [branches/deferred-timeouts-5786-4/twisted/internet/defer.py](branches/deferred-timeouts-5786-4/twisted/internet/defer.py)
    1. "Don't use this. Use L{timeoutDefered}." -- Maybe add a link to a ticket about deprecating then removing this API.

 2. [branches/deferred-timeouts-5786-4/twisted/test/test_defer.py](branches/deferred-timeouts-5786-4/twisted/test/test_defer.py)
    1. "class TimeoutTests(unittest.TestCase):" -- Maybe inherit from SynchronousTestCase instead.
    2. test_noTimeoutIfCallback
       1. The first two assertions seem unnecessary
       2. And the important last assertion might be better tested by checking IDelayedCall.active on the return value of timeoutDeferred.
    3. test_noTimeoutIfErrback
       1. Same applies here.
    4. test_noTimeoutIfCancel
       1. Again I might be nicer to check whether the DelayedCall is still active.
    5. test_timeout
       1. Seems like it would be better to advance the reactor to exactly the timeout value rather than *slightly* before then after.
       2. And again assert that the DelayedCall is not active.
    6. test_callbackStack
       1. I found the docstring difficult to understand...or at least the docstring didn't seem to accurately describe the test implementation.
       2. And I'm not sure this test is necessary since it seems to be testing the standard cancellation behaviour rather than anything specific to timeoutDeferred.
    7. test_multipleTimeouts
       1. Again, isn't this just testing the standard cancellation behaviour of chained deferreds, or am I misunderstanding?
    8. test_cancelReturnedDelayedCall
       1. I'd quite like to see this near the top of the TestCase because it's such an important part of the new API and so that the test story flows neatly into subsequent tests that then make use of the returned DelayedCall.

 3. I don't see a response to glyph's 3D rendering usecase in [[#5786](https://github.com/twisted/twisted/issues/5786)](https://github.com/twisted/twisted/issues/5786):comment:22. His comment seems to be more about the cancellation API in general, but perhaps his point here is that there are circumstances where you need to differentiate between cancellation due to a timeout and some other cancellation reason. How could a user do that if they used this API?

 4. Some of my comments match those made by tom in #5786:comment:26

As far as I'm concerned this can be merged after you answer or address the numbered points above. It's been through 3 rounds of review from glpyh, exarkun and tom.prince. But if you'd like another opinion then resubmit for another review.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_35"></a>|@cyli commented|
|-|-|

Apologies, I'm somewhat late coming into this, but since it's been a while...

We have similar code at https://github.com/rackerlabs/otter/blob/master/otter/util/deferredutils.py#L60 (tests here; https://github.com/rackerlabs/otter/blob/master/otter/test/test_deferredutils.py#L15) that we've been using for about 2 years, and looks and behaves pretty much like Itamar's code, and addresses Glyph's comment about distinguishing between cancellations and timed out errors.

Unless the user passed a cancellation function in in the first place, in which case bets are off.

Perhaps these two pieces of code could be merged to address this ticket?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_36"></a>|@glyph commented|
|-|-|

Replying to [cyli](#note_35):
> Apologies, I'm somewhat late coming into this, but since it's been a while...
> 
> We have similar code at https://github.com/rackerlabs/otter/blob/master/otter/util/deferredutils.py#L60 (tests here; https://github.com/rackerlabs/otter/blob/master/otter/test/test_deferredutils.py#L15) that we've been using for about 2 years, and looks and behaves pretty much like Itamar's code, and addresses Glyph's comment about distinguishing between cancellations and timed out errors.

Partially to clarify for other observers, and partially just to make sure I understand for myself the additional feature that the code from Otter is implementing;

Otter's `timeout_deferred` doesn't attempt what Itamar originally suggested, which is to say, a different exception at the _start_ of the callback chain - you call `cancel()` and the canceler does whatever it does, so the initial callbacks in the chain receive whatever values they expect, including `CancelledError`.

Instead, it adds a callback to the Deferred _at the point in the chain where it is called_, which means that the caller can know what kind of translation they're getting and when they're getting it, and not interfere with the earlier callback chain.

The one issue I have with the code in Otter is that it's not _quite_ flexible enough.  While translating `CancelledError` to `TimeoutError` is in the case of a failure is _usually_ what you want, there are weird edge-cases.  Maybe a partial result is unacceptable, and you really want to fail with a `TimeoutError` unconditionally once you've cancelled.  Maybe you're dealing with a _subclass_ of `CancelledError` that exposes some important information about the progress during cancellation, and naively translating to `TimeoutError` naively would throw that away and that's bad.

So I would say that while this makes a reasonable default, the exception (or result!) translation callable ought to be an optional parameter to `timeoutDeferred`, effectively a callback added via `addBoth`, but one which is only invoked if the timeout has fired.

> Unless the user passed a cancellation function in in the first place, in which case bets are off.

Actually I think that the behavior here is pretty well-defined; if they specified a custom canceler, they can get the behavior that they implemented; with a custom error translator this can behave however the caller wants.

> Perhaps these two pieces of code could be merged to address this ticket?

Yes; looking over Itamar's branch, it seems like it would be very straightforward to add the error-translation stuff from otter in (with the tweak I recommended) and get something that satisfies everything brought up in the discussion on this ticket so far.  Thanks!

**Comment by twisted-trac** [comment]
|<img alt="txmanish's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_38"></a>|txmanish commented|
|-|-|

Hello. Looks like cyli has addressed all the issues which rwall had with otter's implementation. If so, should I submit a patch with that code? You can review it and I can fix any issues raised. I would really like to have this in Twisted as I am looking to use it in other projects too.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_39"></a>|@glyph commented|
|-|-|

Please do.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_41"></a>|@cyli set owner to @cyli|
|-|-|

Stealing for PyCon2016

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_42"></a>|@cyli commented|
|-|-|

https://github.com/twisted/twisted/pull/145

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_45"></a>|@adiroiban set owner to @cyli|
|-|-|

Thanks for the changes.

Major comments:

1. I don't understand the translateCancellation example. I was expecting it to be based on the previous example.

From the translateCancellation docs and API docs I don't understand how I should use the translateCancellation... should I return something from it? How it should know how to handle timeout cancellations ?

So maybe it can be renamed to `onTimeoutCancel` as it will be called when the deferred is cancelled with a timeout. In this way users can do whatever they want here :)

2. Do we need the new twisted.internet.task.TimedOutError ? Can we re-use twisted.internet.error.TimeoutError ?

3. We will be on 16.4.0 next.

4. I think that cancelledToTimedOutError should be _cancelledToTimedOutError to make it explicit a private function.

5. I can not make sense of the last part of the release notes:

```
can produce a TimedOutError, or other custom error, distinct CancelledErrors.
```


Minor comments:

1. In api doc we should have L{int} instead of C{int} to make it explicit that this is a cross reference. Same for C{None}. 

2. I don't think there is any value in using _DummyException in tests... it can be DummyException or mabye give it a less generic name... like SomeGenericException  or CustomCancelException ... the current DummyException is used for both purposes but I think that this does not help making the tests easier to read :)

3. I think that the functionality tested by test_defaultTranslationPreservesCancellationFunctionCallback should be part of the narrative docs ... or if the current documentation already talks about it should be some more obvious that when a custom cancellation is used, the timeout will no longer work.

-------------

Since I don't understand how you would like the translateCancellation to be used, I am not reviewing the remaining tests as I don't know if they are valid or not.


Please note that there is also this ticket [#8533](https://github.com/twisted/twisted/issues/8533) which tries to clean the current mess in twisted.internet.defer regarding the TimeoutError and the timeout() function. For this ticket, I think that you can ignore that work, but maybe it will help to get an idea of what we had in the past.


Please check my current comments and resubmit for review .

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_46"></a>|@cyli removed owner|
|-|-|

Hi Adi!

Thank you for your detailed review.  I think I have addressed all your review comments:

1.  I've renamed the callable timeoutDeferred takes to `onTimeoutCancel` (thanks for the suggestion!)
2.  I've Cleaned up the documentation so that rather than discussing how to "translate custom CancelledErrors to TimedOutErrors" we talk about the `onTimeoutCancel` callable as just a function that gets called when the deferred is timed out, which can be used for logging or returning different values.  This seems like a more reasonable use case.
3.  I totally forgot `twisted.internet.error.TimeoutError` was there, thanks - I've used it instead.
4.  I've fix the tests to be more clear about explicitly testing that timing out doesn't happen if the deferred is callbacked/errbacked/canceled before the timeout, since canceling again can cancel dependent deferreds.
5.  I've bumped the @since tag to 16.5 and updated the topfile to just talk about the function itself rather than all the options it can take or errors it can return.

Thanks!

(https://github.com/twisted/twisted/pull/145)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_47"></a>|@glyph set owner to @cyli|
|-|-|

Thanks cyli!

I think this is almost ready to go; looks like all the existing feedback was dealt with.

There are a couple of minor issues though:

1. In the tradition of the methods on `Deferred`, `timeoutDeferred` should really return its argument.  This would make chaining possible, which can make code that uses this read more nicely.  For example, you could shorten your example in the docs to:
```python
later = task.deferLater(reactor, delay, f)
def called(result):
    print("{0} seconds later:".format(delay), result)
return timeoutDeferred(later, 3, reactor).addBoth(called)
```
2. If I set a timeout on a `Deferred`, but then I *later* decide that I want to cancel or extend the timeout, how would I do that?  The reason I make this blocking feedback is that it probably affects the signature here and I'd like to avoid a deprecation cycle.  However, I'm not sure that it's actually an important thing for the API to support.

Then the non-blocking things:

1. Over the years `twisted.internet.task` has taken on some really unfortunate scope-creep.  The reason it's called `task` is that it was originally the home of `Cooperator`, which allows you to break up a _long-running task_ into small pieces to run cooperatively with the reactor.  However, `Cooperator` needed to run periodically in the reactor, so that gave rise to `LoopingCall`, which lived next to it in the same module.  Later, since `LoopingCall` had a `Deferred` in it, and also had some time-related code, `deferLater` came to live in this module as well.  Now it seems like `task` has become the dumping ground for everything related to any potential intersection between `Deferred` and the passage of time.  The net result here is that we have this confusingly-named module which no longer really has anything to do with "tasks" at all.  The relevance to _this_ change is that it makes me wonder if this functionality should live in `defer.py` after all.  The previous `Deferred.setTimeout` was a disaster which should _not_ have been in `defer.py`, but that was because it imported the reactor directly and therefore tightly bound `defer.py` to the rest of `twisted.internet`.  Given that this method takes an IReactorTime from the outside, it could become a method on `Deferred` without having `defer.py` import anything.  I'm sorry that this review point was quite long-winded; I just wanted to give the necessary background here.  However, I'm going to leave it up to the author's discretion; I think the possibilities should be leaving it where it is, moving it to a top-level function in `defer.py`, or maybe even making it a method on `Deferred`.
2. You don't need to use `C{}` / `L{}` markup in comments; it won't be processed by anything.
3. Rather than using `[0]` for function closeures, I've started getting in the habit of using attributes on the function objects themselves.  So for example, instead of `timedOut = [False]`, something like this instead:
```python
    def timeItOut():
        timeItOut.didTimeOut = True
        deferred.cancel()
    timeItOut.didTimeOut = False
    "..."
    if timeItOut.didTimeOut:
        "..."
```
Totally optional, it just lets you skip using zero as a magic number.  (It is trading in one hack for another; what we _want_ is to use the `nonlocal` keyword, but no dice there until we can drop python 2...)

Thanks again for checking back in on this long-suffering issue :).  I think the first required point is really all you _need_ to take care of here, and that's a sufficiently minor detail (assuming you add a test for it) that I think you can just land it without re-review.  If you want to make any more significant changes based on my optional feedback, you should probably submit for re-review.  Up to you!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_48"></a>|@cyli commented|
|-|-|

Thanks for reviewing, Glyph!

At the San Francisco Twisted meetup, there was a discussion amongst attendees (including glyph, runciter, moshez, and others) about some of these API design discussions.  Summary:

1.  No one particularly objects to moving this to the ```defer``` package.
2.  No one could really think of a good use case for needing to cancel a timeout.  The possible cases that were suggested were better served being implemented using ```LoopingCall```, and perhaps we should document an example of such a case where a `LoopingCall` based implementation would be more appropriate.
3.  Everyone seem to like the idea of ```deferred.addTimeout(...)```, so that one could do something like: ```deferred.addCallbacks(...).addCallback(...).addErrback(...).addTimeout(...).addCallback(...)```

I'll make these changes.  :)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_49"></a>|@cyli removed owner|
|-|-|

Updated - I think maybe the ```LoopingCall``` vs ```addTimeout``` docs should maybe go in a different prose doc ticket?  Maybe something like "timing recipes"?  I'm not sure yet what a good section for that would be - it seems kind of random to say "oh and in this one very special case, use ```LoopingCall}} instead of {{{addTimeout```.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_50"></a>|@markrwilliams commented|
|-|-|

https://github.com/twisted/twisted/pull/145#pullrequestreview-2209020

Feel free to put this back in the review queue.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_51"></a>|@markrwilliams commented|
|-|-|

From GitHub:

The implementation is good and the documentation sufficient.  I would like at least some of the tests I put into [that comment](https://github.com/twisted/twisted/pull/145#discussion_r81226759) included in this PR to elucidate the interaction between timeouts and existing callbacks and errbacks.  They're merely nice to have, though, so if I don't see them here in the next 24 hours I'll merge this and open a new PR to add them.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_52"></a>|@markrwilliams set owner to @markrwilliams<br>@markrwilliams set status to `closed`|
|-|-|

In changeset 63f3e6e5acfc9042cb0a7c925ed321d3814d3241
```
#!CommitTicketReference repository="" revision="63f3e6e5acfc9042cb0a7c925ed321d3814d3241"
Merge pull request #145 from twisted/timeout-deferreds-5786-5

Merge twisted/timeout-deferreds-5786-5: addTimeout function on defer.Deferred

Author: cyli 
Reviewer: adiroiban, theisencouple, manishtomar, markrwilliams
Fixes: #5786

Added twisted.internet.defer.Deferred.addTimeout method to enable timeouts of deferreds.
```
